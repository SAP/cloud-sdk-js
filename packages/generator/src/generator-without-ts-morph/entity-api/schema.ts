import { documentationBlock } from '@sap-cloud-sdk/util';
import { VdmEntity, VdmProperty } from '../../vdm-types';
import { addLeadingNewline, getStaticPropertyDescription } from '../../typedoc';
import { createPropertyFieldInitializerForEntity } from '../../entity';

/**
 * @internal
 */
export function getSchema(entity: VdmEntity): string {
  return `{ 
    ${[
      ...properties(entity),
      '...this.navigationPropertyFields',
      // ...navigationProperties(entity, service),
      allFieldsSelector(entity)
    ].join(',\n')} 
  }`;
}

function properties(entity: VdmEntity): string[] {
  return entity.properties.map(prop => property(prop));
}

function property(prop: VdmProperty): string {
  return [
    documentationBlock`${getStaticPropertyDescription(prop)}`,
    `${prop.staticPropertyName}: ${createPropertyFieldInitializerForEntity(
      prop,
      'fieldBuilder'
    )}`
  ].join('\n');
}

function allFieldsSelector(entity: VdmEntity): string {
  return [
    documentationBlock`${addLeadingNewline('All fields selector.')}`,
    `ALL_FIELDS: new AllFields('*', ${entity.className})`
  ].join('\n');
}
