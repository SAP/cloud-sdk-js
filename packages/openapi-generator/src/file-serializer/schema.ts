import { codeBlock, documentationBlock, unixEOL } from '@sap-cloud-sdk/util';
import {
  OpenApiSchema,
  OpenApiObjectSchema,
  OpenApiObjectSchemaProperty
} from '../openapi-types';
import { getType } from '../parser/type-mapping';
import {
  isReferenceObject,
  isArraySchema,
  isObjectSchema,
  isEnumSchema,
  isOneOfSchema,
  isAllOfSchema,
  isAnyOfSchema,
  isNotSchema,
  getSchemaPropertiesDocumentation
} from '../schema-util';

/**
 * Serialize a schema.
 * @param schema - Parsed schema to be serialized.
 * @returns The serialized schema
 * @internal
 */
export function serializeSchema(schema: OpenApiSchema): string {
  if (isReferenceObject(schema)) {
    return schema.schemaName;
  }
  if (isArraySchema(schema)) {
    const type = serializeSchema(schema.items);
    return schema.uniqueItems ? `Set<${type}>` : `${type}[]`;
  }
  if (isObjectSchema(schema)) {
    return serializeObjectSchema(schema);
  }

  if (isEnumSchema(schema)) {
    return schema.enum.join(' | ');
  }

  if (isOneOfSchema(schema)) {
    return schema.oneOf.map(type => serializeSchema(type)).join(' | ');
  }

  if (isAllOfSchema(schema)) {
    return schema.allOf.map(type => serializeSchema(type)).join(' & ');
  }

  if (isAnyOfSchema(schema)) {
    return schema.anyOf.map(type => serializeSchema(type)).join(' | ');
  }

  if (isNotSchema(schema)) {
    return codeBlock`any`;
  }

  return getType(schema.type);
}

function serializeObjectSchema(schema: OpenApiObjectSchema): string {
  if (
    !schema.properties.length &&
    typeof schema.additionalProperties !== 'object'
  ) {
    return 'Record<string, any>';
  }

  const types: string[] = [];
  if (schema.properties.length) {
    types.push(serializeObjectSchemaForProperties(schema.properties));
  }

  if (schema.additionalProperties) {
    types.push(
      `Record<string, ${serializeSchema(schema.additionalProperties)}>`
    );
  }

  return types.join(' | ');
}

function serializeObjectSchemaForProperties(
  properties: OpenApiObjectSchemaProperty[]
): string {
  return codeBlock`{
      ${properties
        .map(property => serializePropertyWithDocumentation(property))
        .join(unixEOL)}
    }`;
}

function serializeProperty(name: string, required: boolean, type: string) {
  return `'${name}'${required ? '' : '?'}: ${type};`;
}

function serializePropertyWithDocumentation(
  property: OpenApiObjectSchemaProperty
) {
  const documentation = schemaPropertyDocumentation(property);
  const serialized = serializeProperty(
    property.name,
    property.required,
    serializeSchema(property.schema)
  );
  if (documentation) {
    return [documentation, serialized].join(unixEOL);
  }
  return serialized;
}

/**
 * @internal
 */
export function schemaPropertyDocumentation(
  schema: OpenApiObjectSchemaProperty
): string {
  const signature: string[] = [];
  if (schema.description) {
    signature.push(schema.description);
  }

  signature.push(...getSchemaPropertiesDocumentation(schema.schemaProperties));

  return documentationBlock`${signature.join(unixEOL)}`;
}
