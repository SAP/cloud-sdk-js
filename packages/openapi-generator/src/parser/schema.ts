import { createLogger } from '@sap-cloud-sdk/util';
import { OpenAPIV3 } from 'openapi-types';
import { isReferenceObject } from '../schema-util';
import {
  OpenApiArraySchema,
  OpenApiEnumSchema,
  OpenApiObjectSchema,
  OpenApiObjectSchemaProperty,
  OpenApiReferenceSchema,
  OpenApiSchema,
  OpenApiSchemaProperties
} from '../openapi-types';
import { getType } from './type-mapping';
import { OpenApiDocumentRefs } from './refs';
import { ParserOptions } from './options';

const logger = createLogger('openapi-generator');

/**
 * Parse the original schema or reference object to a serializable schema.
 * @param schema - Originally provided schema or reference object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema.
 * @internal
 */
export function parseSchema(
  schema: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiSchema {
  if (!schema) {
    logger.verbose("No schema provided, continuing with 'any'.");
    return { type: 'any' };
  }

  if (isReferenceObject(schema)) {
    return parseReferenceSchema(schema, refs);
  }

  if (schema.type === 'array') {
    return parseArraySchema(schema, refs, options);
  }

  if (
    schema.type === 'object' ||
    schema.properties ||
    'additionalProperties' in schema
  ) {
    return parseObjectSchema(schema, refs, options);
  }

  if (schema.enum?.length) {
    return parseEnumSchema(schema, options);
  }

  if (schema.oneOf?.length) {
    return parseXOfSchema(schema, refs, 'oneOf', options);
  }

  if (schema.allOf?.length) {
    return parseXOfSchema(schema, refs, 'allOf', options);
  }

  if (schema.anyOf?.length) {
    return parseXOfSchema(schema, refs, 'anyOf', options);
  }

  if (schema.not) {
    return {
      not: parseSchema(schema.not, refs, options)
    };
  }

  return {
    type: getType(schema.type)
  };
}

function parseReferenceSchema(
  schema: OpenAPIV3.ReferenceObject,
  refs: OpenApiDocumentRefs
): OpenApiReferenceSchema {
  return {
    ...schema,
    ...refs.getSchemaNaming(schema)
  };
}

/**
 * Parse a schema to an array schema.
 * @param schema - Original schema representing an array.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed array schema.
 */
function parseArraySchema(
  schema: OpenAPIV3.ArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiArraySchema {
  return {
    uniqueItems: schema.uniqueItems,
    items: parseSchema(schema.items, refs, options)
  };
}

/**
 * Parse a schema to an object schema.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed object schema.
 */
function parseObjectSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiObjectSchema {
  const properties = parseObjectSchemaProperties(schema, refs, options);

  if (schema.additionalProperties === false) {
    if (!properties.length) {
      throw new Error(
        'Could not parse object schema without neither properties nor additional properties.'
      );
    }

    return { properties };
  }

  const additionalProperties =
    typeof schema.additionalProperties === 'object' &&
    Object.keys(schema.additionalProperties).length
      ? parseSchema(schema.additionalProperties, refs, options)
      : { type: 'any' };

  return {
    properties,
    additionalProperties
  };
}

/**
 * Parse properties of an object as property schemas.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The list of parsed property schemas.
 */
function parseObjectSchemaProperties(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiObjectSchemaProperty[] {
  return Object.entries(schema.properties || {}).reduce(
    (props, [propName, propSchema]) => [
      ...props,
      {
        schema: parseSchema(propSchema, refs, options),
        description: isReferenceObject(propSchema)
          ? undefined
          : propSchema.description,
        name: propName,
        required: schema.required?.includes(propName) || false,
        schemaProperties: { ...parseSchemaProperties(propSchema) }
      }
    ],
    []
  );
}

/**
 * Parse an enum schema.
 * @param schema - Original schema representing an enum.
 * @param options - Options that were set for service generation.
 * @returns The parsed enum schema.
 */
function parseEnumSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  options: ParserOptions
): OpenApiEnumSchema {
  const type = schema.type ? getType(schema.type) : 'string';
  return {
    type,
    enum: (schema.enum || []).map(entry => {
      if (type === 'string' && entry !== null) {
        return getEnumStringValue(String(entry));
      }
      if (entry === null && !schema.nullable) {
        if (options.strictNaming) {
          throw new Error(
            'null was used as a parameter in an enum, although the schema was not declared as nullable'
          );
        } else {
          logger.warn(
            'null was used as a parameter in an enum, although the schema was not declared as nullable'
          );
        }
      }
      return entry;
    })
  };
}

function getEnumStringValue(input: string): string {
  return `'${input.replace(/'/g, "\\'")}'`;
}

/**
 * Parse a 'oneOf', 'allOf' or 'anyOf' schema.
 * @param schema - Original schema to parse.
 * @param refs - Object representing cross references throughout the document.
 * @param xOf - Key to identify which schema to parse.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema based on the given key.
 */
function parseXOfSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  xOf: 'oneOf' | 'allOf' | 'anyOf',
  options: ParserOptions
): any {
  return {
    [xOf]: (schema[xOf] || []).map(entry => parseSchema(entry, refs, options))
  };
}

/**
 * Parse schema properties e.g. 'maxLength', 'minimum', etc.
 * @param schema - Original schema representing a ref or schema object.
 * @returns The parsed schema properties object.
 * @internal
 */
export function parseSchemaProperties(
  schema: OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject
): OpenApiSchemaProperties {
  if (isReferenceObject(schema)) {
    return {};
  }
  const schemaPropertyNames = [
    'deprecated',
    'example',
    'format',
    'default',
    'multipleOf',
    'maximum',
    'minimum',
    'maxLength',
    'minLength',
    'minItems',
    'maxItems',
    'pattern'
  ];
  return schemaPropertyNames.reduce((properties, propertyName) => {
    if (schema[propertyName]) {
      return { ...properties, [propertyName]: schema[propertyName] };
    }
    return properties;
  }, {});
}
