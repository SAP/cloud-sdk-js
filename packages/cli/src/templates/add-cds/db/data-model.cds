namespace my.bookshop;
using { User, Country, managed } from '@sap/cds/common';

entity Books {
  key ID : Integer;
  title  : localized String;
  author : Association to Authors;
  stock  : Integer;
}

entity Authors {
  key ID : Integer;
  name   : String;
  books  : Association to many Books on books.author = $self;
}

entity Orders : managed {
  key ID  : UUID;
  book    : Association to Books;
  country : Country;
  amount  : Integer;
}