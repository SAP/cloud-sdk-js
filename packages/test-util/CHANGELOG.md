# @sap-cloud-sdk/test-util

## 2.5.0

### Patch Changes

- Updated dependencies [89f1c423]
  - @sap-cloud-sdk/util@2.5.0
  - @sap-cloud-sdk/connectivity@2.5.0

## 2.4.0

### Patch Changes

- Updated dependencies [04726a35]
- Updated dependencies [0a008674]
- Updated dependencies [0a008674]
- Updated dependencies [0a008674]
  - @sap-cloud-sdk/connectivity@2.4.0
  - @sap-cloud-sdk/util@2.4.0
