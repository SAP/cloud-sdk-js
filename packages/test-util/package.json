{
  "name": "@sap-cloud-sdk/test-util",
  "version": "1.32.1",
  "description": "SAP Cloud SDK for JavaScript test utilities",
  "homepage": "https://community.sap.com/topics/cloud-sdk",
  "license": "Apache-2.0",
  "keywords": [
    "sap-cloud-sdk",
    "cloud-sdk",
    "sap-cloud-platform",
    "test-util"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist/**/*.js",
    "dist/**/*.js.map",
    "dist/**/*.d.ts",
    "dist/**/*.d.ts.map"
  ],
  "repository": "github:SAP/cloud-sdk",
  "scripts": {
    "compile": "yarn tsc -b",
    "prepare": "yarn compile",
    "test": "yarn jest --coverage --runInBand",
    "test:local": "yarn jest",
    "test:debug": "node --inspect-brk ../../node_modules/.bin/jest --runInBand",
    "check:dependencies": "depcheck ."
  },
  "dependencies": {
    "@sap-cloud-sdk/core": "^1.32.1",
    "@sap-cloud-sdk/util": "^1.32.1"
  }
}
