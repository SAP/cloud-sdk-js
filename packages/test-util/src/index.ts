/**
 * [[include:test-util/README.md]]
 * @packageDocumentation
 * @module @sap-cloud-sdk/test-util
 */

export * from './test-destination-provider';
export * from './test-destination-mocker';
