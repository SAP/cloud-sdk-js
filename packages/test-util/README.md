<!-- sap-cloud-sdk-logo -->
<!-- This block is inserted by scripts/replace-common-readme.ts. Do not adjust it manually. -->
<a href="https://sap.github.io/cloud-sdk/docs/js/overview"><img src="https://help.sap.com/doc/2324e9c3b28748a4ae2ad08166d77675/1.0/en-US/logo-with-js.svg" alt="SAP Cloud SDK for JavaScript Logo" height="122.92" width="226.773"/></a>
<!-- sap-cloud-sdk-logo-stop -->

# @sap-cloud-sdk/test-util

This package contains utility functions for testing, like loading credentials or creating test destinations.

## Installation

```
$ npm install @sap-cloud-sdk/test-util
```

## Usage

The test-util package makes writing tests for your SAP Business Technology Platform application more convenient.

For example, you can create a mock destination for your tests by using the function `mockTestDestination`.

```
import { mockTestDestination } from '@sap-cloud-sdk/test-util';
mockTestDestination('mockDestinationName');
```

You need to save your system information in local file `systems.json`:

```json
{
  "systems": [
    {
      "alias": "mockDestinationName",
      "uri": "https://www.example.com"
    }
  ]
}
```

and `credentials.json`:

```json
{
  "credentials": [
    {
      "alias": "SYS_001",
      "username": "username",
      "password": "password"
    }
  ]
}
```

<!-- sap-cloud-sdk-common-readme -->
<!-- This block is inserted by scripts/replace-common-readme.ts. Do not adjust it manually. -->
## Support

The recommended way to get in touch with us is to create an issue on [GitHub](https://github.com/SAP/cloud-sdk-js/issues).
Select the issue category **Bug**, **Feature**, or **Question** depending on the nature of your request.
We try to provide fixes, features and answers as soon as possible.

## Contribute

If you would like to contribute to the SAP Cloud SDK, please make yourself familiar with our [contributing guidelines](https://github.com/SAP/cloud-sdk-js/blob/main/CONTRIBUTING.md) and follow the given instructions.

## Links

- [Official support channel](https://github.com/SAP/cloud-sdk-js/issues/new/choose)
- [Github](https://github.com/SAP/cloud-sdk-js)
- [SAP Cloud SDK Documentation portal](https://sap.github.io/cloud-sdk)
  - [SAP Cloud SDK Documentation portal - Overview](https://sap.github.io/cloud-sdk/docs/js/overview)
  - [SAP Cloud SDK Documentation portal - Getting started guide](https://sap.github.io/cloud-sdk/docs/js/getting-started)
  - [SAP Cloud SDK Documentation portal - API documentation](https://sap.github.io/cloud-sdk/api/latest)
  - [SAP Cloud SDK Documentation portal - Release notes](https://sap.github.io/cloud-sdk/docs/js/release-notes)
- [Sample repository](https://github.com/SAP-samples/cloud-sdk-js)

## License

The SAP Cloud SDK is released under the [Apache License Version 2.0.](http://www.apache.org/licenses/)
<!-- sap-cloud-sdk-common-readme-stop -->
