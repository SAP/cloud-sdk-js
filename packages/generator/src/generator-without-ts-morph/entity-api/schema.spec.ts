import { breakfastEntity } from '../../../test/test-util/data-model';
import { getSchema } from './schema';

describe('entity api schema', () => {
  it('gets proper schema', () => {
    expect(getSchema(breakfastEntity)).toMatchSnapshot();
  });
});
