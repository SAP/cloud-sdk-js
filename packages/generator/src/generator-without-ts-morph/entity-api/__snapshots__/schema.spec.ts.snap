// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity api schema gets proper schema 1`] = `
"{ 
    /**
 * Static representation of the {@link entityName} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
ENTITY_NAME: fieldBuilder.buildEdmTypeField('EntityName', 'Edm.String', false),
/**
 * Static representation of the {@link numberOfEggs} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
NUMBER_OF_EGGS: fieldBuilder.buildEdmTypeField('A_noEggs', 'Edm.Decimal', true),
/**
 * Static representation of the {@link breakfastTime} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
BREAKFAST_TIME: fieldBuilder.buildEdmTypeField('BreakfastTime', 'Edm.DateTime', false),
...this.navigationPropertyFields,
/**
 * 
 * All fields selector.
 */
ALL_FIELDS: new AllFields('*', Breakfast) 
  }"
`;
